[RAPTOR-12838] add streaming to chat example

ScottOglesby opened the pull request:
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Small change to add streaming support to the chat example for `custom.py`

## Rationale
* a reference implementation for OpenAI chat streaming
* a convenient streaming stub model for dev and test

## Changes
* `chat()` hook now respects the `stream` field in `completion_create_params`
* each chunk is one word of the response
* default 0.25 second delay between chunks
* response uses OpenAI Python classes
* cleaned up file to satisfy pylint and black

## Testing
```
export TARGET_NAME=completions
drum server --code-dir model_templates/python3_dummy_chat/ --target-type textgeneration --address localhost:6789
```

```
from openai import OpenAI

openai_client = OpenAI(
    base_url="http://localhost:6789",
    api_key="not used",
)

# non-streaming
completion = openai_client.chat.completions.create(
    model="datarobot-deployed-llm",
    messages=[{"role": "user", "content": "What is the meaning of life?"}],
)
print(completion.choices[0].message.content)

# streaming
completion = openai_client.chat.completions.create(
    model="datarobot-deployed-llm",
    messages=[{"role": "user", "content": "What is the meaning of life?"}],
    stream=True,
)
for chunk in completion:
    print(chunk)
```

peterzdeb commented on model_templates/python3_dummy_chat/custom.py:
```diff
@@ -62,29 +60,63 @@ def load_model(code_dir: str) -> Any:
     message_content = "Echo: " + completion_create_params["messages"][0]["content"]
+    stream = completion_create_params.get("stream", False)
+
+    if stream:
+        # Mimic OpenAI streaming: yield one chunk at a time, split by whitespace
+        def gen_chunks():
```
Do we have tests for this?

ScottOglesby replied:
@peterzdeb There generally is "can the custom model be created" testing (`test_inference_model_templates.py`) and some negative tests (`test_other_cases.py`).

However, I created a more direct test in `test_model_templates.py`. It's unit instead of full e2e, but exercises the python model adapter using a Flask server, with an OpenAI client driving the test. It adds some value, and we can add other model template tests later.

ScottOglesby commented:
@yakov-g I added a test for the chat hook based on Petro's review comment. Also moved a few fixtures into `tests/unit/datarobot_drum/drum/conftest.py` since they are now used in 2 test modules.

The PR already has 2 approvals, but the changes are large enough that "sneaking them in" seems wrong. I'll ping for a re-review to make sure the test changes are OK.
